### PR TITLE
Add missing bitwise ops: xor, unary not (~), and shifts

### PR DESCRIPTION
**Pull request by Hittorp**

### CODE REVIEW

**Comment by Hittorp** [comment]
I signed it!

**Comment by jayconrod** [comment]
Hey @Hittorp, thanks for sending this.

@adonovan is offline for a while, so you may want to wait for him to get back for a canonical response on this PR. That said, I don't think we can accept this without some agreement from the Bazel team. These bitwise operators are not simply missing from the implementation; they're not in the language specification. So adding them is a change to the language, not just filling in missing functionality.

If you'd like to move forward with this, start by bringing it up on the [bazel-discuss](https://groups.google.com/forum/#!forum/bazel-discuss) mailing list. If there is interest there, you can [submit a design doc](https://bazel.build/designs/index.html). If that's approved, the change can be implement in Bazel and here.

I expect these operators were deliberately left out of the language though. I'm not sure why exactly, but I expect this is due to the desire to keep Skylark as small as possible. Adding these operators may introduce additional complexity (especially with big integers) without sufficient benefit (Skylark, as a configuration language, isn't used for much bitwise math).

**Comment by laurentlb** [comment]
It's unlikely we add this to Bazel.

There are a few extensions in this repository that are not Bazel (e.g. lambdas, nested functions, floating point numbers). Bitwise operators could be added in a similar way, behind a flag, is we believe they are important enough.

**Comment by Hittorp** [comment]
I have added -bitwiseops flag to enable new ops only if it's set.

Thanks @laurentlb for the suggestion.

**Comment by alandonovan** [comment]
Thanks Hittorp, the change looks very clean, though I've only given it a first pass so far. The only minor things that stand out are:
- the spec should describe the semantics of the new operators
- also, the incompatibilities section should describe how the language differs from the Bazel implementation (new tokens, etc)
- a test case should cover the 1<<toobig case, and any new errors.

Regarding the question of whether this change to the language is a good idea: bitwise operations are a fundamental primitive operation on 2's-complement integers and have been supported in hardware by every ALU since the IBM 704 in the mid-1950s. They are crucial to interpreting bitfields such as appear in UNIX file modes and many other domains that one might wish to use a configuration language. The syntax, semantics, and implementation are obvious and uncontroversial, and follow Python, and the implementation imposes no cost on those that do not need the feature. So, I support adding them to Skylark. However, if the Bazel team does not wish to add support in their implementation, then the logic must live behind a dialect flag, as you have it in this change.

**Comment by Hittorp** [comment]
Hi Alan, I have changed spec.md as you requested and put 1<<toobig test to int.sky. Previous commits were amended. 
After that the Travis CI is failed by some reason on Go:master with:
> FAIL: ExampleThread_Load_parallelCycle (0.00s)
got:
cannot load a.sky: cannot load c.sky: cannot load b.sky: cycle in load graph
cannot load b.sky: cycle in load graph
want:
cannot load a.sky: cannot load c.sky: cycle in load graph
cannot load b.sky: cannot load a.sky: cannot load c.sky: cycle in load graph

Could you please suggest what's happening? Go:1.9.x build looks good and also my local 'go test' too. I have not touched the load() implementation.

**Comment by alandonovan** [comment]
Thanks Hittorp, don't worry about the test, it's been flaky for a long
while and I haven't gotten around to fixing it.

I can't log in to GitHub at this minute because of a 2-factor problem, so
we'll have to use e-mail:

- typo: "This features" -> "These features"
- please add the << and >> operators to the precedence table under "Binary
operators"
- please add and explain << etc in the Arithmetic operations section
- the existing & and | operators are bitwise ops on integers and also set
operations on sets.
  Could you add a similar check to the resolver to disallow & and | unless
the -bitwise option is specified?
  (This check is imperfect because & and | are also set operations, but
since the check operates on the syntax, we can only check for operators. I
don't really want to add a dynamic dialect check.)
  don't forget to update the comment at AllowBitwiseOps.
  The three resolver error message "support xor and shift bitwise
operations" can all be just "support bitwise operations"
- I notice I failed to add &= and |= when I added & and |.  If you're
feeling diligent, you could add them too.
- in the interp switch case syntax.LTLT, syntax.GTGT, let's just call
AsInt32 and add a negative check.  I don't think we need AsUint32 for a
single call.
- also there: use an 'else' block to emphasize the symmetry:
    if op == syntax.LTLT {
        return x.Lsh(y), nil
    } else {
        return x.Rsh(y), nil
    }
- ^ should have higher precedence than |, following Python. Please update
parser and spec, and test.


Many thanks again.

**Comment by alandonovan** [comment]
One last thing: we should probably issue a dynamic error on x << y where y
is too large, to avoid eating all available memory.
The Go compiler's constant computation reports "shift count too large: %d"
unless y < 512, so let's do that too.

**Comment by Hittorp** [comment]
Alan, I've made suggested changes. Could you please review them?

File: cmd/skylark/skylark.go [hunk 1]
```diff
@@ -32,6 +32,7 @@ func init() {
 	flag.BoolVar(&resolve.AllowSet, "set", resolve.AllowSet, "allow set data type")
 	flag.BoolVar(&resolve.AllowLambda, "lambda", resolve.AllowLambda, "allow lambda expressions")
 	flag.BoolVar(&resolve.AllowNestedDef, "nesteddef", resolve.AllowNestedDef, "allow nested def statements")
+	flag.BoolVar(&resolve.AllowBitwise, "bitwise", resolve.AllowBitwise, "allow bitwise operations (&, |, ^, unary ~, <<, and >>)")
```
Comment (adonovan) [inline cmd/skylark/skylark.go]: ~ is always unary, so you can drop the adjective.

File: doc/spec.md [hunk 2]
```diff
@@ -834,6 +842,10 @@ applied to sets.  The right operand of the `|` operator may be any
 iterable value.  The binary `in` operator performs a set membership
 test when its right operand is a set.
 
+The binary `^` operator performs symmetric difference of two sets.
+The resulting set will have an elements of either first or second
```
Comment (adonovan) [inline doc/spec.md]: typo: "an elements"

But I think you can drop the second sentence (The resulting...) as this section is just an overview of the available operators, not their semantics in detail.

File: doc/spec.md [hunk 3]
```diff
@@ -1876,12 +1908,25 @@ The result of `set | set` is a new set whose elements are the
...
 operands, left before right.
 
+The `^` operator accepts operands of either `int` or `set` type.
+For integers, it yields the bitwise XOR (exclusive OR) of its operands.
+For sets, it yields a new set containing elements of either first or second
+operand but not both (symmetric difference).
+
+The `<<` and `>>` operators require operands of `int` type both. They shift
+the first argument to the left or right by the number of bits given by the
+second operand.
```
Comment (adonovan) [inline doc/spec.md]: "It is a dynamic error if the second argument is negative. Implementations may impose a limit on the second operand of a left shift."

File: doc/spec.md [hunk 4]
```diff
@@ -1876,12 +1908,25 @@ The result of `set | set` is a new set whose elements are the
 union of the operands, preserving the order of the elements of the
 operands, left before right.
 
+The `^` operator accepts operands of either `int` or `set` type.
+For integers, it yields the bitwise XOR (exclusive OR) of its operands.
+For sets, it yields a new set containing elements of either first or second
+operand but not both (symmetric difference).
+
+The `<<` and `>>` operators require operands of `int` type both. They shift
+the first argument to the left or right by the number of bits given by the
```
Comment (adonovan) [inline doc/spec.md]: s/argument/operand/

File: doc/spec.md [hunk 5]
```diff
@@ -2421,11 +2466,11 @@ In the Java implementation, targets cannot be dot expressions.
 
 An augmented assignment, which has the form `lhs op= rhs` updates the
 variable `lhs` by applying a binary arithmetic operator `op` (one of
-`+`, `-`, `*`, `/`, `//`, `%`) to the previous value of `lhs` and the value
+`+`, `-`, `*`, `/`, `//`, `%`, `^`, `<<`, `>>`) to the previous value of `lhs` and the value
```
Comment (adonovan) [inline doc/spec.md]: and &= and |= too

(strike "arithmetic" in the previous line)

File: eval.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ import (
 )
 
 const debug = false
+const maxShiftCount = 511
```
Comment (adonovan) [inline eval.go]: Move this constant to where it is needed.  It doesn't need a name.

File: resolve/resolve.go [hunk 7]
```diff
@@ -90,6 +90,7 @@ var (
 	AllowFloat          = false // allow floating point literals, the 'float' built-in, and x / y
 	AllowSet            = false // allow the 'set' built-in
 	AllowGlobalReassign = false // allow reassignment to globals declared in same file (deprecated)
+	AllowBitwise        = false // allow bitwise operations (&, |, ^, unary ~, <<, and >>)
```
Comment (adonovan) [inline resolve/resolve.go]: strike "unary"

**Review by adonovan — COMMENTED** [review]
Nearly there---looking good.

**Comment by adonovan** [comment]
Just curious: what are you using Skylark for?

**Comment by Hittorp** [comment]
Hi Alan, I've updated the commits once again. Thank you for the response.
We've been looking for a Go-implemented Python dialect for experimenting within an internal project and Skylark happend to be the best fit. We much appreciate your work on Skylark and ready to promptly provide more details when the project becomes public, if you like.

**Comment by alandonovan** [comment]
Many thanks!  Do send more details about your project when you can.

**Pull request merged**